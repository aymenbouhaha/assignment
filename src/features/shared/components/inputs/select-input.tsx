import { cn } from "@/lib/utils";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@shared/components/ui";
import { ChevronDown } from "lucide-react";
import { Icon } from "@shared/components/icons/icons.tsx";
import { useEffect, useState } from "react";

interface SelectInputProps {
	placeholder: string;
	values: { placeholder: string; value: string }[];
	inputClassName?: string;
	contentClassName?: string;
	onValueChange?: (value: string) => void;
	defaultValue?: string;
}

export const SelectInput = ({
	placeholder,
	values,
	inputClassName,
	contentClassName,
	onValueChange,
	defaultValue = "",
}: SelectInputProps) => {
	const [open, setOpen] = useState(false);
	const [value, setValue] = useState<string>("");

	useEffect(() => {
		if (onValueChange) onValueChange(value);
	}, [value]);

	return (
		<div className="w-full h-full">
			<Select onValueChange={setValue} defaultValue={defaultValue} value={value} open={open} onOpenChange={setOpen}>
				<div className="relative w-full">
					<SelectTrigger
						className={cn(inputClassName, {
							"text-grey-1": !value,
							"text-primary-black": value,
						})}
					>
						<SelectValue placeholder={placeholder} />
					</SelectTrigger>
					{value ? (
						<Icon.Close
							className={"stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"}
							onClick={() => setValue("")}
						/>
					) : (
						<ChevronDown
							className="stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"
							onClick={() => setOpen(true)}
						/>
					)}
				</div>
				<SelectContent className={cn(contentClassName)}>
					{values.map((item, index) => {
						return (
							<SelectItem key={`${item.value}.${index}`} value={item.value}>
								{item.placeholder}
							</SelectItem>
						);
					})}
				</SelectContent>
			</Select>
		</div>
	);
};
