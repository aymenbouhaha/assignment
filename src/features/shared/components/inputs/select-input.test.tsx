import { render, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import { SelectInput } from "@shared/components/inputs/select-input.tsx";

describe("SelectInput", () => {
	const mockValues = [
		{ placeholder: "Option 1", value: "value1" },
		{ placeholder: "Option 2", value: "value2" },
	];
	const mockOnValueChange = vi.fn();

	it("renders correctly with placeholder and no selected value", () => {
		const { asFragment } = render(
			<SelectInput placeholder="Select an option" values={mockValues} onValueChange={mockOnValueChange} />,
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders correctly with a selected value", () => {
		const { getByText, asFragment } = render(
			<SelectInput placeholder="Select an option" values={mockValues} onValueChange={mockOnValueChange} />,
		);

		fireEvent.click(getByText("Select an option"));
		fireEvent.click(getByText("Option 1"));

		expect(mockOnValueChange).toHaveBeenCalledWith("value1");
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders correctly with the clear icon when a value is selected", () => {
		const { getByText, container, asFragment } = render(
			<SelectInput placeholder="Select an option" values={mockValues} onValueChange={mockOnValueChange} />,
		);

		fireEvent.click(getByText("Select an option"));
		fireEvent.click(getByText("Option 2"));

		const clearIcon = container.querySelector(".size-6.cursor-pointer");
		expect(clearIcon).toBeTruthy();

		fireEvent.click(clearIcon!);
		expect(mockOnValueChange).toHaveBeenCalledWith("");

		expect(asFragment()).toMatchSnapshot();
	});
});
