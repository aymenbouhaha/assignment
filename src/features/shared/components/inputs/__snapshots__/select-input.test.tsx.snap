// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectInput > renders correctly with a selected value 1`] = `
<DocumentFragment>
  <div
    class="w-full h-full"
  >
    <div
      class="relative w-full"
    >
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r3:"
        aria-expanded="false"
        class="flex w-full items-center justify-between whitespace-nowrap rounded-md border border-grey-border focus:border-grey-2 bg-[#fff] px-3 py-2.5 ring-0 !text-p21 lg:!text-p3 outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 text-primary-black"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Option 1
        </span>
      </button>
      <svg
        class="stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"
        fill="none"
        height="24"
        stroke="none"
        viewBox="0 0 25 24"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.5 7L7.5 17M7.5 7L17.5 17"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SelectInput > renders correctly with placeholder and no selected value 1`] = `
<DocumentFragment>
  <div
    class="w-full h-full"
  >
    <div
      class="relative w-full"
    >
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r0:"
        aria-expanded="false"
        class="flex w-full items-center justify-between whitespace-nowrap rounded-md border border-grey-border focus:border-grey-2 bg-[#fff] px-3 py-2.5 ring-0 !text-p21 lg:!text-p3 outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 text-grey-1"
        data-placeholder=""
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Select an option
        </span>
      </button>
      <svg
        class="lucide lucide-chevron-down stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SelectInput > renders correctly with the clear icon when a value is selected 1`] = `
<DocumentFragment>
  <div
    class="w-full h-full"
  >
    <div
      class="relative w-full"
    >
      <button
        aria-autocomplete="none"
        aria-controls="radix-:ra:"
        aria-expanded="false"
        class="flex w-full items-center justify-between whitespace-nowrap rounded-md border border-grey-border focus:border-grey-2 bg-[#fff] px-3 py-2.5 ring-0 !text-p21 lg:!text-p3 outline-none disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 text-grey-1"
        data-placeholder=""
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          Select an option
        </span>
      </button>
      <svg
        class="lucide lucide-chevron-down stroke-grey-2 absolute size-6 top-1/2 -translate-y-1/2 right-3 cursor-pointer"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
